src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2181 @@
+            env(batch::batch(alice, seq, batchFee, tfAllOrNothing),
+                batch::add(ammCreate(XRP(10), USD(10)), seq + 1),
+                batch::add(pay(alice, bob, XRP(10)), seq + 2),
+                ter(telINSUF_FEE_P));
+            env.close();
+
+            BEAST_EXPECT(env.seq(alice) == seq);
+            BEAST_EXPECT(env.balance(alice) == preAlice);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+        }
+    }
+
+    void
+    testChangesBetweenViews(FeatureBitset features)
+    {
+        testcase("changes between views");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const carol = Account("carol");
+        env.fund(XRP(220), alice, bob, carol);
+        env.close();
+
+        auto const preAlice = env.balance(alice);
+        auto const preBob = env.balance(bob);
+
+        // Using 1 XRP to create insufficient reserve result
+        auto const batchFee = XRP(1);
+        auto const seq = env.seq(alice);
+        env(batch::batch(alice, seq, batchFee, tfAllOrNothing),
+            batch::add(pay(alice, bob, XRP(10)), seq + 1),
+            batch::add(pay(alice, bob, XRP(10)), seq + 2),
+            ter(tesSUCCESS));
+        TxID const batchId = env.tx()->getTransactionID();
+        std::vector<TestBatchData> testCases = {};
+        env.close();
+        validateBatch(env, batchId, testCases);
+
+        // Alice pays fee and sequence; Bob should not be affected.
+        BEAST_EXPECT(env.seq(alice) == seq + 1);
+        BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+        BEAST_EXPECT(env.balance(bob) == preBob);
+    }
+
+    void
+    testBadInnerFee(FeatureBitset features)
+    {
+        testcase("bad inner fee");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+        XRPAmount const feeDrops = env.current()->fees().base;
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const carol = Account("carol");
+        auto const gw = Account("gw");
+        auto const USD = gw["USD"];
+
+        env.fund(XRP(1000), alice, bob, carol, gw);
+        env.close();
+        env.trust(USD(1000), alice, bob);
+        env(pay(gw, alice, USD(100)));
+        env(pay(gw, bob, USD(100)));
+        env.close();
+
+        auto const preAlice = env.balance(alice);
+        auto const preBob = env.balance(bob);
+
+        auto const seq = env.seq(alice);
+        auto const batchFee = calcBatchFee(env, 0, 2);
+        auto tx = pay(alice, bob, XRP(1000));
+        tx[jss::Fee] = to_string(feeDrops);
+        env(batch::batch(alice, seq, batchFee, tfAllOrNothing),
+            batch::add(pay(alice, bob, XRP(10)), seq + 1),
+            batch::add(tx, seq + 2),
+            ter(tesSUCCESS));
```

Comment:
Shouldn't this test fail with `temINVALID_FEE` or something?

Reply:
Batch::Preflight should catch this...